### ISSUE TITLE
Use TimeSpan everywhere we use an int for seconds, milliseconds, and timeouts

### ISSUE THREAD

**Issue by briangru** [dotnet/runtime#14336]
In a large software project, it's easy to mess up units for time.  IE, is that Int32 seconds or milliseconds?  Or maybe minutes?  I've been fixing my team's source code to use the unit-agnostic TimeSpan class wherever possible.  However the .NET Framework is not complete in its adoption of TimeSpan.  Specifically, we don't have a Process.WaitForExit overload that takes a TimeSpan, only an Int32 for the timeout.  

I suggest someone look through all .NET Framework API's for Int32 parameters containing "second", "millisecond", "ms", "timeout" (and perhaps "time") and see if there is a parallel TimeSpan-based overload.  If not, please fix that.

### API Proposal

*([Proposed](https://github.com/dotnet/runtime/issues/14336#issuecomment-483874118) by @reflectronic)*

```diff
namespace System {
    public static class GC {
+       public static GCNotificationStatus WaitForFullGCApproach(TimeSpan timeout);
+       public static GCNotificationStatus WaitForFullGCComplete(TimeSpan timeout);
    }
}

namespace System.ComponentModel.DataAnnotations {
    public class RegularExpressionAttribute {
!       This one might not end up being that useful, since people generally don't ever manipulate an instance of this attribute.
+       public TimeSpan MatchTimeout { get; }
    }
}

namespace System.Diagnostics {
    public class Process {
+       public bool WaitForExit(TimeSpan timeout);
+       public bool WaitForInputIdle(TimeSpan timeout);
    }
}

namespace System.IO {
    public class FileSystemWatcher {
+       public WaitForChangedResult WaitForChanged(WatcherChangeTypes changeType, TimeSpan timeout);
    }

    public sealed class NamedPipeClientStream : PipeStream {
+       public void Connect(TimeSpan timeout);
+       public Task ConnectAsync(TimeSpan timeout, CancellationToken cancellationToken = default);
    }
}

namespace System.Net.NetworkInformation {
    public class Ping {
+       public PingReply Send(IPAddress address, TimeSpan timeout, byte[]? buffer = null, PingOptions? options = null);
+       public PingReply Send(string hostNameOrAddress, TimeSpan timeout, byte[]? buffer = null, PingOptions? options = null);

!       Skipped EAP based methods; if they are desired, they can be added back in

!       I added CancellationToken because it is probably worth it. If you don't want it you can remove it.
+       public Task<PingReply> SendPingAsync(IPAddress address, TimeSpan timeout, byte[]? buffer = null, PingOptions? options = null, CancellationToken cancellationToken = default);
+       public Task<PingReply> SendPingAsync(string hostNameOrAddress, TimeSpan timeout, byte[]? buffer = null, PingOptions? options = null, CancellationToken cancellationToken = default);
    }
}

namespace System.Net.Sockets {
    public class NetworkStream : Stream {
+       public void Close(TimeSpan timeout);
    }

    public class Socket {
+       public bool Poll(TimeSpan timeout, SelectMode mode);
+       public static void Select(IList checkRead, IList checkWrite, IList checkError, TimeSpan timeout);
    }
}

namespace System.ServiceProcess {
    public class ServiceBase {
+       public void RequestAdditionalTime(TimeSpan time);
    }
}

namespace System.Threading.Tasks {
    public class Task {
+       public bool Wait(TimeSpan timeout, CancellationToken cancellationToken); 
    } 
} 

namespace System.Timers {
    public class Timer {
+       public Timer(TimeSpan interval);
    }
}
```

**APIs that I don't know how to make better:**

```diff
namespace System.IO {
    public abstract class Stream {
+       public TimeSpan ReadTimeoutTimeSpan { get; set; }
+       public TimeSpan WriteTimeoutTimeSpan { get; set; }
    }
}

namespace System.IO.Ports {
    public class SerialPort {
+       public TimeSpan ReadTimeoutTimeSpan { get; set; }
+       public TimeSpan WriteTimeoutTimeSpan { get; set; }
    }
}

namespace System.Media {
    public class SoundPlayer {
+       public TimeSpan LoadTimeoutTimeSpan { get; set; } 
    }
}

namespace System.Net.NetworkInformation {
+   public static class NetworkInformationTimeSpanExtensions {
+       public static TimeSpan GetPacketReassemblyTimeout(this IPGlobalStatistics statistics);
+       public static TimeSpan GetMaximumTransmissionTimeout(this TcpStatistics statistics);
+       public static TimeSpan GetMinimumTransmissionTimeout(this TcpStatistics statistics);
+   }
}

namespace System.Net.Sockets {
    public class Socket {
+       public TimeSpan ReceiveTimeoutTimeSpan { get; set; }
+       public TimeSpan ReceiveTimeoutTimeSpan { get; set; }
    }

    public class TcpClient {
+       public TimeSpan ReceiveTimeoutTimeSpan { get; set; }
+       public TimeSpan SendTimeoutTimeSpan { get; set; }
    }
}

namespace System.Timers {
    public class Timer {
+       public TimeSpan IntervalTimeSpan { get; set; }
    }
}
```

**APIs which use seconds:**

These suffer from the same problem as the previous group, but also only deal with seconds, meaning that most TimeSpan values would need to be rounded out.

```diff
namespace System.Data {
!   This would need to be a DIM.
    public interface IDbCommand {
+       TimeSpan CommandTimeoutTimeSpan { get; set; }
    }

+   public static class DataTimeSpanExtensions {
+       public static TimeSpan GetConnectionTimeout(this IDbConnection connection);
+   }

}

namespace System.Data.Sql {
    public sealed class SqlNotificationRequest {
+       public TimeSpan TimeoutTimeSpan { get; set; }
    }
}

namespace System.Data.SqlClient {
    public sealed class SqlBulkCopy {
+       public TimeSpan BulkCopyTimeoutTimeSpan { get; set; }
    }

    public sealed class SqlConnectionStringBuilder {
+       public TimeSpan ConnectTimeoutTimeSpan { get; set; }
+       public TimeSpan LoadBalanceTimeoutTimeSpan { get; set; }
    }

    public sealed class SqlDependency {
+       public SqlDependency(SqlCommand command, string options, TimeSpan timeout);
    }
}

namespace System.Net.Sockets {
    public class LingerOption {
+       public LingerOption(bool enable, TimeSpan time);
+       public TimeSpan LingerTimeSpan { get; set; }
    }

    public class Socket {
+       public void Close(TimeSpan timeout);
    }
}
```

**Comment by joperezr** [comment]
@reflectronic thanks for providing the list, that is very handy.

@terrajobst regarding your comment:
>There is a lot of trickery due to .NET Framework compat and how we type forward contracts.

Is there anything impeading us adding these overloads just in netcoreapp? In case that's fine, then I guess this issue might be ready to take to API Review.

**Comment by terrajobst** [comment]
LOL. No idea what I meant. It was a different time back then. Maybe I though how we can OOB these methods? Who knows. Let’s just look at the APIs in one of next reviews. Seems generally like a fine feature.

**Comment by joperezr** [comment]
sounds good, will mark it as ready for review then.

**Comment by reflectronic** [comment]
Er... there wasn't actually an API proposal before. However, I have now updated my original comment with an API proposal.

**Comment by GSPP** [comment]
@reflectronic It seems you are proposing adding parallel properties. So both an `int` version and a `TimeSpan` version. What behavior would result from reading and writing both properties? When I set 1.5 seconds using the `TimeSpan` version, what do I get reading the `int` version? Probably a rounded result.

Thanks to default interface methods it might actually be feasible to modify the interfaces.

I think it's quite problematic from an API design standpoint to have both versions. An alternative would be to use `TimeSpan` for any new timeout APIs but keep existing ones using integers.

Another alternative is to really put `TimeSpan` versions *everywhere* (in parallel to the old ones), deprecate the old ones, create a Roslyn based automatic migration tool (this should just work for 99% of cases) and delete the `int` versions in the next major version. This would be really awesome if this was feasible.

Using `int` for a time span in the first place is a pretty obvious design bug stemming from .NET 1.0. It would be awesome if we could totally get rid of that.

The concept of having an automatic API upgrade tool (based on Roslyn) should be explored more. Not just for this situation. Maybe .NET 5 is the right time to pull the trigger on some cleanups like this one.

**Comment by aevitas** [comment]
@GSPP What would you envision such an upgrade tool to do? Would Roslyn analyzers with code fixes that allow people to migrate from the old overloads to the modern `TimeSpan` variants be feasible or do you think an API migration requires a more "rigid" approach?

Could be interesting to explore this path and see just how feasible providing (semi-)automated migrations from older APIs would be.

**Comment by GSPP** [comment]
@aevitas I'm not too familiar with the Roslyn analyzer infrastructure. These upgrades need to happen in mass, across the entire solution, with a preview and a report of the changed locations.

If the old APIs are to be dropped in .NET 5 then opening a pre-5 solution with .NET 5 would immediately result in a lot of compiler errors. The tool should then go and fix them.

Maybe .NET 5 needs to keep all old APIs and delete them in .NET 6.

In the long term, I think it is very undesirable to keep both versions. Either the old version must eventually be deleted, or the change not taken at all, IMHO.

**Comment by terrajobst** [comment]
Related: https://github.com/dotnet/runtime/issues/30723

**Comment by terrajobst** [comment]
* We're happy to approve the first group, i.e. a set of overloads to existing methods where `int` is replaced with `TimeSpan`.
* We don't believe properties or using DIMs for interfaces is worth it.

```diff
namespace System {
    public static class GC {
+       public static GCNotificationStatus WaitForFullGCApproach(TimeSpan timeout);
+       public static GCNotificationStatus WaitForFullGCComplete(TimeSpan timeout);
    }
}

namespace System.ComponentModel.DataAnnotations {
    public class RegularExpressionAttribute {
!       This one might not end up being that useful, since people generally don't ever manipulate an instance of this attribute.
+       public TimeSpan MatchTimeout { get; }
    }
}

namespace System.Diagnostics {
    public class Process {
+       public bool WaitForExit(TimeSpan timeout);
+       public bool WaitForInputIdle(TimeSpan timeout);
    }
}

namespace System.IO {
    public class FileSystemWatcher {
+       public WaitForChangedResult WaitForChanged(WatcherChangeTypes changeType, TimeSpan timeout);
    }

    public sealed class NamedPipeClientStream : PipeStream {
+       public void Connect(TimeSpan timeout);
+       public Task ConnectAsync(TimeSpan timeout, CancellationToken cancellationToken = default);
    }
}

namespace System.Net.NetworkInformation {
    public class Ping {
+       public PingReply Send(IPAddress address, TimeSpan timeout, byte[]? buffer = null, PingOptions? options = null);
+       public PingReply Send(string hostNameOrAddress, TimeSpan timeout, byte[]? buffer = null, PingOptions? options = null);

!       Skipped EAP based methods; if they are desired, they can be added back in

!       I added CancellationToken because it is probably worth it. If you don't want it you can remove it.
+       public Task<PingReply> SendPingAsync(IPAddress address, TimeSpan timeout, byte[]? buffer = null, PingOptions? options = null, CancellationToken cancellationToken = default);
+       public Task<PingReply> SendPingAsync(string hostNameOrAddress, TimeSpan timeout, byte[]? buffer = null, PingOptions? options = null, CancellationToken cancellationToken = default);
    }
}

namespace System.Net.Sockets {
    public class NetworkStream : Stream {
+       public void Close(TimeSpan timeout);
    }

    public class Socket {
+       public bool Poll(TimeSpan timeout, SelectMode mode);
+       public static void Select(IList checkRead, IList checkWrite, IList checkError, TimeSpan timeout);
    }
}

namespace System.ServiceProcess {
    public class ServiceBase {
+       public void RequestAdditionalTime(TimeSpan time);
    }
}

namespace System.Threading.Tasks {
    public class Task {
+       public bool Wait(TimeSpan timeout, CancellationToken cancellationToken); 
    } 
} 

namespace System.Timers {
    public class Timer {
+       public Timer(TimeSpan interval);
    }
}
```

**APIs that I don't know how to make better:**

```diff
namespace System.IO {
    public abstract class Stream {
+       public TimeSpan ReadTimeoutTimeSpan { get; set; }
+       public TimeSpan WriteTimeoutTimeSpan { get; set; }
    }
}

namespace System.IO.Ports {
    public class SerialPort {
+       public TimeSpan ReadTimeoutTimeSpan { get; set; }
+       public TimeSpan WriteTimeoutTimeSpan { get; set; }
    }
}

namespace System.Media {
    public class SoundPlayer {
+       public TimeSpan LoadTimeoutTimeSpan { get; set; } 
    }
}

namespace System.Net.NetworkInformation {
+   public static class NetworkInformationTimeSpanExtensions {
+       public static TimeSpan GetPacketReassemblyTimeout(this IPGlobalStatistics statistics);
+       public static TimeSpan GetMaximumTransmissionTimeout(this TcpStatistics statistics);
+       public static TimeSpan GetMinimumTransmissionTimeout(this TcpStatistics statistics);
+   }
}

namespace System.Net.Sockets {
    public class Socket {
+       public TimeSpan ReceiveTimeoutTimeSpan { get; set; }
+       public TimeSpan ReceiveTimeoutTimeSpan { get; set; }
    }

    public class TcpClient {
+       public TimeSpan ReceiveTimeoutTimeSpan { get; set; }
+       public TimeSpan SendTimeoutTimeSpan { get; set; }
    }
}

namespace System.Timers {
    public class Timer {
+       public TimeSpan IntervalTimeSpan { get; set; }
    }
}
```

**Comment by GrabYourPitchforks** [comment]
Might recommend nixing `System.Timers` from the list. That's largely a dead API surface.

**Comment by danmosemsft** [comment]
Not essential to ship 5.0

**Comment by huoyaoyuan** [comment]
Open question:
Some apis use -1 for special meaning (infinity). How to represent it in `TimeSpan`?
For input cases, we can just disable the infinity semantic. But for output (namely `RegularExpressionAttribute.MatchTimeout`), I can't find a suitable representation.

**Comment by huoyaoyuan** [comment]
Contribution pain:

The implementation for these apis are simple enough, but the major pain should be documentation, since the policy that new api must have xml doc in source. Some of these don't have xmldoc, some have outdated ones.

Options for contribution:
- Wait for the doc cleanup epic to complete
- Omit the policy for now, and add these new apis into the cleanup epic. They should be just a few rewording from `int` overloads.

**Comment by adamsitnik** [comment]
@huoyaoyuan I think that you can just add new methods with xml docs and don't worry about the existing methods that don't have the xml docs, as they are going to be added as part of separate work. Please let me know if you need any help.

**Comment by mattjohnsonpint** [comment]
> Some apis use -1 for special meaning (infinity). How to represent it in `TimeSpan`?

`TimeSpan.MaxValue` is a little over 29,227 *years*.  I think in most cases that is good enough.

**Comment by jogibear9988** [comment]
> > Some apis use -1 for special meaning (infinity). How to represent it in `TimeSpan`?
> 
> `TimeSpan.MaxValue` is a little over 29,227 _years_. I think in most cases that is good enough.
maybe. some part of the code would setup. a callback of a value other than infinity, so this could be a perf. consern

**Comment by drieseng** [comment]
@jogibear9988 There are already APIs that expect a **TimeSpan** representing **-1** milliseconds to indicate infinity.
One such example is [`System.Threading.Thread.Sleep(TimeSpan timeout)`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.thread.sleep#System_Threading_Thread_Sleep_System_TimeSpan_).

As @huoyaoyuan mentioned, this value is currently exposed through [`System.Threading.Timer.InfiniteTimeSpan`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.timeout.infinitetimespan).

**Comment by danmoseley** [comment]
Assigning to @deeprobin

**Comment by danmoseley** [comment]
Hmm, @deeprobin I can't assign to you unless you're either an org member (which you aren't) or you've commented on the issue. If you comment here, I can assign it.

**Comment by deeprobin** [comment]
@danmoseley 
Oh thanks for the info Dan. Here is my comment.

I did apply to the Foundation a while back, but I don't know at what intervals they look at their applications :smile:.

**Comment by deeprobin** [comment]
@danmoseley 
I think we could mark the issue as `api-suggestion` and let the 2nd group review already.

Currently only the 1st group is approved (See https://github.com/dotnet/runtime/issues/14336#issuecomment-636135854). Then I can take care of the other groups in a follow-up PR.

**Comment by danmoseley** [comment]
> I did apply to the Foundation a while back, but I don't know at what intervals they look at their applications

That's different -- I'm talking about a Github security group that merely gives you the right to be assigned without a comment. If this keeps coming up, we can add you, but most contributors aren't in it.

> I think we could mark the issue as api-suggestion and let the 2nd group review already.

I suspect that would become confusing. I suggest to open a new  api suggestion issue for that 2nd group and copy over the relevant info. Then your PR can close this one.

**Comment by bartonjs** [comment]
> I suggest to open a new api suggestion issue for that 2nd group

Yes, please :smile:.

**Comment by danmoseley** [comment]
@deeprobin we need issue(s) for anything not reviewed.
